### PR TITLE
[Frontend] Support PyTrees as function arguments and return values

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
This PR updates the Frontend to support PyTrees as function arguments and return values. 

JAX uses some tree-like data structure for container-like and nested Python structures, such as lists of lists, and dicts of arrays. These structures are the so-called PyTrees in JAX. Catalyst currently doesn’t support JAX PyTrees.

**Benefits:**
- Catalyst supports PyTree as arguments and return values.
- QJIT programs efficiently work with PyTree arguments and return values.

**Requirements:**
Support return values in
- [X] Quantum workflows (with `QNode`)
- [X] Classical workflows (without `QNode`)
- [X] Hybrid workflows
- [X] Measurement results (that expands the original PyTreeDef)
- [X] Quantum/Classical/Hybrid gradient calls
- [X] measure and control-flow calls

Support function arguments in
- [X] Quantum workflows
- [X] Uses type promotion to avoid recompilation
- [X] Quantum/Classical/Hybrid gradient calls
- [X] control-flow calls
- [x] measure
- [x] PyTrees that have `aux` static data

[sc-42673]

### CODE REVIEW

**Review by erick-xanadu — APPROVED** [review]
Looks good! I just deleted a comment. :)

**Comment by maliasadi** [comment]
Great teamwork! :tada:

File: frontend/catalyst/pennylane_extensions.py [hunk 1]
```diff
@@ -85,6 +85,11 @@ def _trace_quantum_tape(
     qreg = qargs[0]
     return_values, qreg, qubit_states = trace_quantum_tape(tape, qreg, has_tracer_return_values)
     qreg = insert_to_qreg(qubit_states, qreg)
+
+    # To support retvals in nested loops
+    if return_values and len(return_values) == 1:
+        return_values = return_values[0]
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:91]: What is the special case being considered here? :)

File: frontend/test/pytest/test_jit_behaviour.py [hunk 2]
```diff
@@ -574,30 +610,12 @@ def addi(x: int, y: int):
...
-    def test_array_repr_from_context(self, coeffs, backend):
-        """Test array representation from context in Hamiltonian."""
-
-        @qjit(target="mlir")
-        @qml.qnode(qml.device(backend, wires=6))
-        def f():
-            qml.Hadamard(wires=0)
-            qml.CNOT(wires=[0, 1])
-            obs = [qml.PauliX(0) @ qml.PauliZ(1), qml.Hadamard(0)]
-            return qml.expval(qml.Hamiltonian(coeffs, obs))
```
Comment (dime10) [inline frontend/test/pytest/test_jit_behaviour.py:594]: Why did we remove this one?

File: doc/changelog.md [hunk 3]
```diff
@@ -20,6 +20,43 @@
   workflow()
   ```
 
+* Add support for PyTrees as function arguments and return values.
```
Comment (maliasadi) [inline doc/changelog.md:23]:
```suggestion
* Add support for container-like and nested Python structures, such as lists of lists or dictionaries
  that can be occurred in a QJIT program as function arguments or return values via JAX ``pytree`` API.
```

**Review by dime10 — APPROVED** [review]
Looks really great 🥳 💯 

The only thing I would double check is to make sure that it also works with the JAX integration since I didn't see any changes in that part of the code.

File: demos/adaptive_circuits_demo.ipynb [hunk 4]
```diff
@@ -897,10 +915,14 @@
     "            apply_selected)\n",
     "\n",
     "num_gates = len(doubles)\n",
-    "params = [0.0] * num_gates\n",
-    "gates_verifier = [True] * num_gates\n",
-    "grads = avqe_circuit_grad(num_gates, 1, params, [0.0], doubles, singles,\n",
-    "            [[-1,-1,-1,-1]], [[-1,-1]], gates_verifier, [False], False)\n",
+    "params = jnp.asarray([0.0] * num_gates)\n",
```
Comment (josh146) [inline demos/adaptive_circuits_demo.ipynb:918]: @maliasadi I'm curious about the change here?

File: doc/changelog.md [hunk 5]
```diff
@@ -72,15 +109,33 @@
 
 <h3>Breaking changes</h3>
 
+* Since we are now using PyTrees, python lists are no longer automatically converted into JAX
+  arrays. This means that indexing on lists when the index is not static will cause a
+  ``TracerIntegerConversionError``.
```
Comment (josh146) [inline doc/changelog.md:114]: Ah, this answers my question above :)

File: doc/changelog.md [hunk 6]
```diff
@@ -72,15 +109,33 @@
...
 
+* Since we are now using PyTrees, python lists are no longer automatically converted into JAX
+  arrays. This means that indexing on lists when the index is not static will cause a
+  ``TracerIntegerConversionError``.
+
+  ```python
+  @qjit
+  def f(list, index):
+    return list[index]
+  ```
```
Comment (josh146) [inline doc/changelog.md:120]: Oh, is this not supported in pure jax?

File: demos/adaptive_circuits_demo.ipynb (hunk 4 above)
Comment (maliasadi) [reply demos/adaptive_circuits_demo.ipynb:918]: We used to convert Python lists to `jax.numpy.array` before tracing in the frontend as JAX doesn't support Python lists. Now, we have support for PyTrees without any implicit list (or array) conversions under the hood. This makes the UI more consistent with JAX's behaivour.

File: doc/changelog.md (hunk 6 above)
Comment (maliasadi) [reply doc/changelog.md:121]: No, it's not supported there either.

File: frontend/catalyst/pennylane_extensions.py (hunk 1 above)
Comment (maliasadi) [reply frontend/catalyst/pennylane_extensions.py:91]: This is for dealing with return values in loops. For example, the `test_nested_loops` in the [test_loops.py](https://github.com/PennyLaneAI/catalyst/blob/39e8611b3491f4e9a32a1b32865da5da732d45e9/frontend/test/pytest/test_loops.py#L223) would be broken without this patch. There may be a better way to fix this issue, but it's working now :)

File: frontend/test/pytest/test_jit_behaviour.py (hunk 2 above)
Comment (maliasadi) [reply frontend/test/pytest/test_jit_behaviour.py]: Oops! Fixed [9d6d2b1](https://github.com/PennyLaneAI/catalyst/pull/221/commits/9d6d2b191949b0d84f66defc7cef70f60d83f172)

File: frontend/catalyst/pennylane_extensions.py (hunk 1 above)
Comment (erick-xanadu) [reply frontend/catalyst/pennylane_extensions.py:91]: I think this is for all primitives that have `multiple_results = True`. As this specifies that the primitive must return either a list of a tuple.

**Review by maliasadi — COMMENTED** [review]
:tada:
